docs/language-reference/07-declarations.md

```diff
@@ -0,0 +1,771 @@
+The mechanisms for parameter passing are specific to each target platform.
+
+> Note: Currently only global shader parameters of opaque types or arrays of opaque types are supported.
+
+A global shader parameter may include an initial-value epxression, but such an expression does not affect the semantics of the compiled program.
+
+> Note: Initial-value expressions on global shader parameters are only useful to set up "default values" that can be read via reflection information and used by application code.
+
+### Variables at Function Scope
+
+Variables declared at _function scope_ (in the body of a function, initializer, subscript acessor, etc.) may be either a function-scope constant, function-scope static variable, or a local variable.
+
+#### Function-Scope Constants
+
+A variable declared at function scope and marked with both `static` and `const` is a _function-scope constant_.
+Semantically, a function-scope constant behaves like a global constant except that is name is only visible in the local scope.
+
+#### Function-Scope Static Variables
+
+A variable declared at function scoep and marked with `static` (but not `const`) is a _function-scope static variable_.
+Semantically, a function-scope static variable behaves like a global static variable except that its name is only visible in the local scope.
+
+The initial-value expression for a function-scope static variable may refer to non-static variables in the body of the function.
+In these cases initialization of the variable is guaranteed not to occur until at least the first time the function body is evaluated for a given invocation.
+
+#### Local Variables
+
+A variable declared at function scope and not marke with `static` (even if marked with `const`) is a _local variable_.
+A local variable has unique storage for each _activation_ of a function by an invocation.
+When a function is called recursively, each call produces a distinct activation with its own copies of local variables.
+
+Functions
+---------
+
+Functions are declared using the `func` keyword:
+
+```hlsl
+func add(x: int, y: float) -> float { return float(x) + y; }
+```
+
+Parameters
+----------
+
+The parameters of the function are declared as `name: type` pairs.
+
+Parameters may be given a _default value_ by including an initial-value-expression clause:
+
+```hlsl
+func add(x: int, y: float = 1.0f) { ... }
+```
+
+Parameters may be marked with a _direction_ which affects how data is passed between caller and callee:
+
+```hlsl
+func add(x: in out int, y : float) { x += ... }
+```
+
+The available directions are:
+
+* `in` (the default) indicates typical pass-by-value (copy-in) semantics. The callee receives a *copy* of the argument passed by the caller.
+
+* `out` indicates copy-out semantics. The callee writes to the parameter and then a copy of that value is assigned to the argument of the caller after the call returns.
```

Comment:
Might need some clarity as to how that differs from say C++.

Reply:
I think the stuff below helps to clarify that `out` and `in out` *may* be implemented using C++ references and their semantics, but that this is not a guarantee provided by the language.